### ISSUE TITLE
Send a 304 response for a just-gone-stale hitpass object when appropriate

### ISSUE THREAD

**Issue by slimhazard** [varnishcache/varnish-cache#2042]
When the hit-for-pass TTL for an uncacheable object expires (that is, on the first request after Varnish stops passing, so that we have a cache miss and go to fetch ), the backend might send a response with an ETag header that matches If-None-Match in the client request. In that case, there's no reason why the client response shouldn't be 304. But evidently resp.status is always 200 in this case.

Presumably the same thing happens when req.http.If-Modified-Since < resp.http.Last-Modified just after the TTL elapses for a hitpass object (but I haven't tried it).
## Expected Behavior

We set beresp.uncacheable to true and beresp.ttl > 0 in vcl_backend_response -- meaning we set the object to hit-for-pass.

After the beresp.ttl elapses, suppose that the backend sets the ETag header, which in turn sets resp.http.ETag, and suppose further that its value matches req.http.If-None-Match. Then I expect client response status 304.
## Current Behavior

Under these conditions (beresp.ttl just elapsed for a hitpass object, and req.http.If-None-Match==resp.http.ETag), Varnish apparently always sets resp.status to 200, and sends the response body.

I'll post a gist with a VTC to illustrate what happens.

Note that while beresp.ttl is unelapsed, so that the object is still in hitpass, the If-None-Match header is passed along to the backend, and if the backend responds with 304 (because INM matches ETag), then status 304 is passed along to the client. The effect is that clients get 304 and no response body most of the time, but suddenly get 200 at random-seeming intervals, despite INM matching ETag.
#1858 fixed the backend request, so that INM and IMS are not sent to the backend for just-elapsed hitpass objects. That's correct, and it may lead to the fetch after the miss getting a 200 response from the backend along with the beresp body. But there's no reason not to set the client response to 304 when the conditions in the client request and response allow it.
## Possible Solution

The 304 client response is set by this code at the end of cnt_deliver:

```
        if (!(req->objcore->flags & OC_F_PASS)
            && req->esi_level == 0
            && http_IsStatus(req->resp, 200)
            && req->http->conds && RFC2616_Do_Cond(req))
                http_PutResponse(req->resp, "HTTP/1.1", 304, NULL);
```

At the moment I don't see why the 304 response isn't set. After the hitpass-TTL elapses, the OC_F_PASS should not be set, so that shouldn't be what's preventing it. Maybe req->http->conds is false?
## Steps to Reproduce (for bugs)

Run the VTC from gist that I will post in a moment.
## Context

We're seeing this for a response with about 700MB in the body, which takes about 6 minutes to transfer, so it is quite costly to send the entire response with status 200. We really want to send 304 to the client whenever we can, and our clients and backend apps are co-operating by setting INM and ETag appropriately.
## Your Environment
- Version used: 4.0.3 and 4.1.2
- Operating System and version: RHEL 6.8 (Linux 2.6.32-642.3.1.el6.x86_64) and Debian 8.5 (Linux 3.16.0-4-amd64)

**Comment by slimhazard** [comment]
VTC posted here:
https://gist.github.com/slimhazard/cf41fcdba1f23e7d7016be66b1ad2a8a

**Comment by fgsch** [comment]
Varnish will not pass the INM for lookups, only if it's a pass.

Are you saying it should or it should check the response and use the INM from the client?

**Comment by slimhazard** [comment]
It's correct not to pass the INM to the bakcned, but nevertheless, the backend may send an ETag in the response. The ETag beresp header becomes the ETag client response header at delivery.

What matters is what we have in the client request and response at delivery time -- that's all that's evaluated in cnt_deliver in the snippet above. When req.http.If-None-Match matches resp.http.ETag, it's justified to set the client response status to 304 and leave out the response body, and essential to performance for very large responses.

It's also peculiar to keep getting 304's as long as hitpass still holds, but then 200 for one request when the TTL expires, even though nothing has changed about INM and ETag, and then get 304's again when the hitpass TTL starts over.

The problem is not what the backend side does, it's what we do on the client side at delivery.

**Comment by fgsch** [comment]
If you remove the `beresp.uncacheble = true` from your example for the second request it will work.
That is converting it to a pass so the `if (!(req->objcore->flags & OC_F_PASS)` will fail.
Can't think of a reason why we restrict this for lookups though.

**Comment by slimhazard** [comment]
Here's another VTC to illustrate what I'm talking about:
https://gist.github.com/slimhazard/d394ebf91d12a9d92a3bf31b3c65b9b4

This more closely matches our real-world use case -- we set beresp.uncacheable to true (and hence go to hit-for-pass) when beresp.ttl <= 0. Overall, we set TTLs almost exclusively based on what's in Cache-Control, so that developers decide whether responses are cached, and for how long. When Cache-Control sets TTL to 0, we go to hit-for-pass -- with few exceptions, this is the only way to know that we can pass. This is why not setting beresp.uncacheable is not an option.

The first two requests/responses do what I'd expect them to do, and so does the third backend request/response. The problem is in the third client request/response -- INM matches ETag, so the client response should be 304.

**Comment by fgsch** [comment]
Right. As I mentioned, the condition is false since the object is a (hit-for-)pass (as opposed to a regular pass).  Dropping that condition fixes it.

Related to #1206. I'm not entirely sure the rationale for that though.

**Comment by slimhazard** [comment]
OK, I think I see now why this happens.

The OC_F_PASS flag gets set in fetch when bo->uncacheable is true (cache_fetch.c in vbf_beresp2obj()) -- this is why http_PutResponse(304) in the snippet above never gets executed. The condition that 304 doesn't get set in delivery was added in the fix for #1858.

Also, http_filterfields() in cache_http.c removes If-None-Match (and If-Modified-Since) in the fetch case (tbl/http_headers.h has an F but not a P in the filter column for both headers), even if  those headers were sent by the client. (Since there's no P, the headers aren't filtered in the pass case.)

So if VCL sets the beresp back to uncacheable after the miss when the hitpass-TTL expires, as we have to do for the use case, then Varnish can never decide to set 304 in delivery. And since we remove the conditional request headers in a fetch (unlike pass), we can never pass along a 304 response from the backend either (which does happen in pass).

I'm beginning to think that the fix for #1858 went too far. The problem there was that the backend may have said 304 and sent no response body, but then delivery might have set the status to 200, and yet there's no response body. We don't want that, but the opposite case is not a problem -- the backend may have sent 200 and a response body, but on the client side we can set status 304, if the conditions are right for it.

**Comment by slimhazard** [comment]
VCL workaround:
https://gist.github.com/slimhazard/5d9b862797c3e3b4d77bab9e071a70a0

Just have vcl_deliver check the conditions from INM/ETag-based 304, and set resp.status accordingly. Presumably the same can be done for IMS/Last-Modified, but it would require parsing the dates and converting them to numerics for comparison.

I would nevertheless argue that varnishd is not doing the right thing in this case. I see two possible approaches to a fix.

One would be to undo the fix for #1206, which of course would be problematic due to the fact that Varnish has been working that way for four years, so that withdrawing it might break existing installations. If we ignore that and only consider "the right thing", then IMO the problem in #1206 should not have been solved in the binary -- *that* should have been a matter for VCL. If a backend declines to set status 304, but nevertheless keeps the values in ETag and/or Last-Modified so that 304 appears to be justified, I would argue that the backend is doing the wrong thing. IMO the solution should have been for someone who's having that problem to remove INM/IMS from bereq in VCL for the pass case. (The idea behind IMS/INM conditional requests in Varnish has always been that VCL authors can choose to remove the headers if they don't want it.)

Since the fix for #1206 has in fact been around for a long time, the other option would be, as fgs has suggested, to distinguish "real pass" (indicated by the OC_F_PASS flag in delivery) from transactions that started out as a miss/fetch but then were changed to pass by setting beresp.uncacheable in VCL (for example by adding a new OC_F_* flag). In those cases, the INM/IMS headers were removed because of miss/fetch, but now we don't do the 304 check in delivery because of the OC_F_PASS flag, meaning that neither the backend or client side ever had a chance to check if 304 is possible. The fix would be: change the guard against the 304 check shown in the snippet above to check both OC_F_* flags, and don't do it only if OC_F_PASS is set but the new flag is not set.

**Comment by slimhazard** [comment]
Yet another workaround, this time also checking IMS and Last-Modified in vcl_deliver:
https://gist.github.com/slimhazard/d2c6b3ee24857160cb15bf9055bc9e13

The part that's a bit clumsy is using `now` as the fallback in `std.time()`, because we have to have something, anything of VCL data type TIME. The fallbacks are set up so that the check fails if either of IMS or Last-Modified are missing.

**Comment by slimhazard** [comment]
As agreed in bugwash -- since 5.0 is a major version upgrade, we can revert the fix for #1206. Will be accompanied with documentation for VCL authors that IMS/INM can be removed if conditional requests are not desired (for example, if you don't want it for pass).

**Comment by nigoroll** [comment]
@slimhazard has discussed, I'd also think a commented-out unset for INM/IMS in builtin.vcl could be helpful

**Comment by nigoroll** [comment]
ftr this relates to #1799 in that
* this issue is related to the hit-for-pass vs. hit-for-miss discussion https://www.varnish-cache.org/lists/pipermail/varnish-dev/2016-September/008989.html
* for both cases we need to solve miss from hit

**Comment by slimhazard** [comment]
Labeling this for bugwash. pre-5.0 hit-for-pass can now be done again (with ``return(pass(DURATION))`` in vcl_backend_response), so this gets me back to the issue that got me started.

This gist illustrates what happens in current trunk:

https://gist.github.com/slimhazard/d3002c147bc285857928b8be65a38b3d

* The backend always sends an ETag (the same considerations will apply for IMS/Last-Modified)
* Response gets set to "classic" hit-for-pass
* The next request after the hfp TTL elapses is a miss, and the client requests sends INM for the same ETag.
* Backend sends a response, again setting the same ETag.

So at that point, req.http.If-None-Match==resp.http.ETag, which ordinarily warrants setting the 304 client response in delivery. But because of the fix in #1206, the 304 client response is not set.

Looking back over this issue, we seem to have already discussed two solutions, both of which would be possible now.

The solution I advocated (and what appears to have been the consensus in previous bugwashes):
* Roll back the fix in #1206 (on the grounds that the fix should have been done in VCL)
* Document the change, and what you'd have to do in VCL for the situation that #1206 wanted to solve (we'll have to add docs for the new/old hfp anyway)
* Add commented-out unsets for IMS/INM to builtin.vcl as guidance for VCL authors.

But fgs' suggestion (in my comment of August 19) is also possible, since in fact there is now a distinct flag ``OC_F_HFP``. So the guard against setting 304 in client delivery would be: don't do it if ``OC_F_PASS`` is set and ``OC_F_HFP`` is not set.

My quick take:
* The first solution is more like "the right thing", if we don't worry about breaking someone's VCL.
* The second solution probably won't break anyone's VCL, but is less like Varnish doing the right thing.

**Comment by slimhazard** [comment]
Decision in today's bugwash -- I'll cook up patches for both solutions, then we'l decide which one we like.

**Comment by slimhazard** [comment]
PRs #2234 and #2235 are the two candidate fixes -- we'll need one or the other of them (not both).

The changes in both of them are just one-liners, so there's not a lot of code to contemplate. It's more about which consequences we prefer, and there's about 750 lines of vtc code to explore that.

PR #2234: Rolls back the fix for #1206, so that client delivery decides whether to set resp.status=304 solely based on req and resp headers, regardless of the PASS state of the fetch. ``resp.status==304`` at ESI level 0 iff ``resp.http.ETag == req.http.If-None-Match`` or ``resp.http.Last-Modified <= req.http.If-Modified-Since``.

PR #2235: Leaves #1206 unchanged, but does not make the exception in delivery if the fetch had been hit-for-pass (as recently revived for 5.1). That is, the decision to set ``resp.status=304`` is considered if ``OC_F_HFP`` was set, but not if ``OC_F_PASS`` was set.

The critical test is what happens if the backend sees the INM/IMS headers in one of the pass cases, decides not to set beresp.status=304, but nevertheless sets ETag or Last-Modified in such a way that it would appear to client delivery that resp.status==304 is warranted. #1206 (and hence PR #2235) took the position that Varnish should do what the backend said to do; PR #2234 takes the position that the backend is in error when it does that, so if you need a fix for a screwy backend, it should be addressed in VCL rather than in varnishd.

At any rate, we didn't want to have the solution in #1206 for the new/old hit-for-pass, because:
* while the hfp TTL is unexpired, the backend sees any INM and IMS the client may have sent, can then set ``beresp.status=304``, and the client sees ``resp.status=304``.
* but when the hfp TTL expires, we have a miss, so INM and IMS are removed from the bereq, so the backend doesn't get a chance to send 304
* for that client response, the client sees ``resp.status=200`` and a (potentially very large) response body, even though the client request and response headers would allow for 304, and the client gets a lot of POLA (especially if the backend would have sent 304 if it could)

For and against PR #2234:
* Behavior is consistent for all variations of pass:
  * "ordinary" pass (``return(pass)`` from ``vcl_recv``)
  * ``beresp.uncacheable=true`` ("hit-for-miss")
  * new/old hit-for-pass (``return(pass(DURATION))`` from ``vcl_backend_response``)
* no POLA effects for conditional requests from the client and status 304
* could break VCL that depends on the existing behavior from the fix for #1206
  * ways to restore the behavior from #1206 in VCL (by unsetting INM/IMS in either vcl_miss or vcl_pass) are demonstrated in ``r02042.vtc`` in PR #2234
  * (I just think that VCL should have been the solution for #1206 all along)
* we'll have to document the possible breakage and the VCL solutions in the 5.1 release (but we'll need to document the return of hfp anyway)

For and against PR #2235:
* the behavior for hfp is not consistent with the other two (but the other two are consistent with each other)
* POLA effects are still there for ordinary pass and hit-for-miss (client request and response headers permit ``resp.status=304``, but the status is 200 and the potentially large response body gets sent)
* won't break anyone's VCL, because no previous release had the new syntax for hit-for-pass, and the other two pass variations are unchanged
* a workaround is possible in VCL, if you prefer the behavior of #2234, by checking the conditions for 304 in vcl_delivery, as shown in the gist mentioned above
  * basically repeats what varnishd would have done in cnt_deliver
  * the workaround is clumsy for IMS/Last-Modified

**Comment by fgsch** [comment]
It might be worth also considering how we treat other request headers in this context, e.g. Range.

**Comment by fgsch** [comment]
The other question we should ask ourselves is whether Varnish should outsmart the backend, or we can/should assume that the backend knows more than us for uncacheable objects.

**Comment by Dridi** [comment]
The sad truth is that we should, and often can't.

**Comment by slimhazard** [comment]
@fgsch what do you have in mind about the other request headers?

**Comment by slimhazard** [comment]
At the risk of getting too philosophical, this is in part about what we expect VCL to be with respect to Varnish and the backends. "varnishd will do what it thinks best, and if it turns out that your backend is being a bit dumb, VCL can fix things up for you" vs. "varnishd will go along with your backend, even if it's a bit dumb, but VCL can cover for that".

More pragmatically, one of the reasons I lean toward rolling back #1206 is that compensating for the change in VCL appears easier -- unset the If-* headers before the fetch. (Which I've always seen as the right approach about the conditional backend fetches: if you don't want them, remove the headers in VCL.) Whereas if we leave #1206 as it is, the workaround is evaluating the client-side headers in vcl_delivery. That's straightforward for ETag/INM, but messy for Last-Modified/IMS.

But I also have to admit that it may be easy for me to say, because I don't think that any of _my_ projects will have a problem with the potential breaking change.

**Comment by nigoroll** [comment]
+1 for #2234 because of consistency & POLY

**Comment by fgsch** [comment]
Considering we also apply ranges for pass'd objects, probably makes sense apply INM/IMS as well.
+1 for #2234 .
